src/nodes.coffee

```diff
@@ -1894,6 +1897,48 @@ exports.Obj = class Obj extends Base
       answer.push @makeCode join
     if @front then @wrapInParentheses answer else answer
 
+  expandProperty: (property) ->
+    {variable, context, operatorToken} = property
+    key = if property instanceof Assign and context is 'object'
+      variable
+    else if property instanceof Assign
+      operatorToken.error "unexpected #{operatorToken.value}" unless @lhs
+      variable
+    else
+      property
+    if key instanceof Value and key.hasProperties()
+      key.error 'invalid object key' unless context isnt 'object' and key.this
+      return new Assign(key, property, 'object', shorthand: yes).withLocationDataFrom property
+    return property unless key is property
+    return property if property instanceof Splat
+
+    new Assign(property, property, 'object', shorthand: yes).withLocationDataFrom property
+
+  expandProperties: ->
+    @expandProperty(property) for property in @properties
+
+  getPropertyAst: (property) ->
+    return property.ast() if property instanceof Splat
+
+    {variable, value, shorthand} = property
+    isComputedPropertyName = variable instanceof Value and variable.base instanceof ComputedPropertyName
+
+    Object.assign
+      type: 'ObjectProperty'
```

Comment:
You'll see this as a code smell. And I agree. In theory I guess this could be delegated to `Assign`'s AST generation (which we haven't gotten to yet). But that feels overloaded/messy and would seem to require making `Assign`s more self-aware of whether they're in fact an object "assignment" or not

To me this screams "missing abstraction"/new node class(/subclass). The logic that I tried to encapsulate in `expandProperty()`/`expandProperties()` to convert the various types that may be in an `Obj`'s `@properties` into a more normalized form (and the similar logic in `Obj::compileNode()`) combined with then having to overload `Assign` for object properties would presumably all disappear if there was a separate node-class-level abstraction for object properties. What is the upside of the existing reuse of `Assign` for object properties?

But (a) I'm not sure exactly how/where to introduce that new abstraction (b) that would be a major refactor, not sure it's worth trying to introduce now?

Reply:
This reminds me of when I was working on `import`/`export`, and I created a `ModuleDeclaration` class to be the parent of `ImportDeclaration` and `ExportDeclaration`. Here we could have something like `BaseAssign` and then `Assign` and `ObjectAssign`. At first `BaseAssign` would have everything that `Assign` does now and `Assign` just extends it without adding anything, and `ObjectAssign` has the extra code specific to objects. I would get that working first, then try to separate out code that only `Assign` needs and `ObjectAssign` doesn’t and gradually move all the “only needed by one of the children but not the other” code out from the base and into the children. I think it’s worth some experimentation.

Reply:
@GeoffreyBooth ok starting to think through this, the biggest thing that makes me question whether subclassing `Assign` is the right abstraction is that there are object properties that aren't assignments/`Assign`s (eg shorthand properties (`{a, [b], @c}`) or splats)

But I should look at `Assign`'s implementation and see what it might take to extract the object-property stuff out completely